### ISSUE TITLE
memory usage

### ISSUE THREAD

**Issue by TojNALvViV** [alacritty/alacritty#5438]
What is alacritty's RSS expected to be?

If I open a new terminal and run top, RSS is 54MB, but if I then hold ctrl- (top still running) until min font size is reached, then hit ctrl+ to restore font size, it's 165MB, and faffing about with zoom like this for a couple of minutes brought it to 375MB.

### System

OS: Linux Ubuntu 20.04LTS x64
Version: alacritty 0.10.0-dev (62a1fad5)
Linux/BSD: X11 dwm

**Comment by chrisduerr** [comment]
Zooming in and out leaking a little bit of memory is a known issue iirc, though I can't find another bug report for it right now.

Though it should take quite a bit of effort to gain any amount of significant memory using that.

**Comment by TojNALvViV** [comment]
It should be pretty easy to get RSS to hundreds of MB, though I haven't been able to get it past 1GB.

One thing I've noticed is that the memory usage appears proportional - at all times - to the area (in pixels) of the terminal. If I run alacritty full-screen (3840x~2400) RSS is above 300MB after just a couple of seconds of zooming. Previously alacritty was about 1920x2400 pixels.

I guess it may be difficult to see this issue in small windows.

**Comment by chrisduerr** [comment]
It's definitely a bug and I can reproduce it no problem. It's just not something that should cause any problems unless you're trying to force the issue.

**Comment by jturcotte** [comment]
I've been having issues when using 100k lines of history where a long running terminal would eventually start taking over 700-1000MB. The strange thing was that the memory would be kept in use even after clearing the terminal.

One issue I've noticed is that `split_off` when shrinking the terminal is not reducing the capacity of every line's `Vec` at https://github.com/alacritty/alacritty/blob/f90dd12/alacritty_terminal/src/grid/row.rs#L78
This is memory that only stays allocated until the terminal is cleared, but could contribute to the high memory usage for the full-screen terminal scenario mentioned here.

The worst cause of the issue for me is however memory fragmentation caused by continuously re-allocating the rows when resizing a terminal containing many history lines.
For this I've locally switched to `tikv-jemallocator` as the global allocator, which handles memory fragmentation better than the default allocator on Linux, and now the terminal usually stays under 100-200MB even with a large history.

I understand that switching the allocator or forcing a reallocation of shrunk lines are possibly not interesting changes if they only benefits scrollback gloutons like me, but I thought that this could be interesting for somebody and worth mentioning here.

**Comment by chrisduerr** [comment]
> The strange thing was that the memory would be kept in use even after clearing the terminal.

That's not super strange really. Generally you don't have much control over forcefully returning memory to the OS. We try to clear our history size if it was shrunk excessively, but that doesn't necessarily mean you get your RAM back.

> One issue I've noticed is that split_off when shrinking the terminal is not reducing the capacity of every line's `Vec`

That's not an issue, that's intended. All lines are always initialized fully, otherwise doing so dynamically would be terrible for performance and make fragmentation even worse.

> The worst cause of the issue for me is however memory fragmentation

It is true that Alacritty can struggle with memory fragmentation, however I don't think there's a simple fix to that. One would have to significantly restructure the underlying data structures to fix that.

**Comment by jturcotte** [comment]
> That's not super strange really. Generally you don't have much control over forcefully returning memory to the OS. We try to clear our history size if it was shrunk excessively, but that doesn't necessarily mean you get your RAM back.

I wouldn't expect 100% of it to be returned, but having a terminal process with an heap usage of 50MB while keeping 700MB booked from the OS, even though expected, is still frustrating for an application that is usually run as multiple long-lived instances on a system.

> It is true that Alacritty can struggle with memory fragmentation, however I don't think there's a simple fix to that. One would have to significantly restructure the underlying data structures to fix that.

I understand. I just wanted to mention that jemalloc does help a lot in my case and is easy to enable on platforms where it's supported.

**Comment by chrisduerr** [comment]
> I understand. I just wanted to mention that jemalloc does help a lot in my case and is easy to enable on platforms where it's supported.

Jemalloc was tested when Rust switched the default away from it and back then it had measurably worse performance. So I don't think it's the best choice.

**Comment by correabuscar** [comment]
> The worst cause of the issue for me is however memory fragmentation caused by continuously re-allocating the rows when resizing a terminal containing many history lines. For this I've locally switched to `tikv-jemallocator` as the global allocator, which handles memory fragmentation better than the default allocator on Linux, and now the terminal usually stays under 100-200MB even with a large history. [source linkback](https://github.com/alacritty/alacritty/issues/5438#issuecomment-1109845572)

@jturcotte How did you do this? (allocator switching)
I've looked at your fork but can't find anything about this in any of the branches.


Related to this issue:
What I've noticed is that a window resize causes the memory usage to double and stay there.
Both VIRT and RES in `top` show this doubling.
Best way to see this is by using a high (say 6 million) history size:
`.config/alacritty/alacritty.yml`:
```yaml
scrolling:
  # Maximum number of lines in the scrollback buffer.
  # Specifying '0' will disable scrolling.
  #history: 10000
  #max is 100k
  #history: 100000
  history: 6000000 #patched

window:
  # Window dimensions (changes require restart)
  #
  # Number of lines/columns (not pixels) in the terminal. Both lines and columns
  # must be non-zero for this to take effect. The number of columns must be at
  # least `2`, while using a value of `0` for columns and lines will fall back
  # to the window manager's recommended size
  dimensions:
    #columns: 80
    columns: 120
    lines: 51
```

```patch
Index: /var/tmp/portage/x11-terms/alacritty-0.11.0/work/alacritty-0.11.0/alacritty_terminal/src/config/scrolling.rs
===================================================================
--- alacritty-0.11.0/alacritty_terminal/src/config/scrolling.rs
+++ alacritty-0.11.0/alacritty_terminal/src/config/scrolling.rs
@@ -4,7 +4,9 @@ use serde::{Deserialize, Deserializer};
 use alacritty_config_derive::{ConfigDeserialize, SerdeReplace};
 
 /// Maximum scrollback amount configurable.
-pub const MAX_SCROLLBACK_LINES: u32 = 100_000;
+pub const MAX_SCROLLBACK_LINES: u32 = 6_000_000; // this 6mil, uses 17g VIRT and 16g RES (in 'top') and
+                                                 // double that after a window resize, and doesn't
+                                                 // go back down!
 
 /// Struct for scrolling related settings.
 #[derive(ConfigDeserialize, Copy, Clone, Debug, PartialEq, Eq)]
```

and then doing this 4 times in a row `time printf '%s %s %s %s %s %s %s %s %s %s\n' {1..15000000}` in bash, which adds 1.5million lines on each run.

Note that I'm not complaining that the above 6 million history lines use 17 gigabytes of memory (I'm fine with that), but rather than the window resizing doubles it to 33g (+-) and it is not temporary (ie. the memory usage remains there at 33g even after the resize is done, doesn't go back down to 17g)
I'm curious is switching the allocator would make any difference.

**Comment by kchibisov** [comment]
One cell takes `24` bytes, so if you multiple by the amount of cell you have you'll get roughly that amount.

The reason it stays is due to memory fragmentation as it was shown by tools like `bytehound`, so the allocator does matter here. The only way to solve this is to change allocation strategy or the allocator.

Rewriting this part of code could be an interesting experience, since you need to preserve performance somehow. The only way I see to solve fragmentation is to have separate vec of indices which you rotate and a separate storage you point to. Adding new vectors to it could be really slow and require a resize for example. Reflow/resize will be a whole new experience as well...

I guess you can use system jemalloc and retest.

**Comment by jturcotte** [comment]
> @jturcotte How did you do this? (allocator switching) I've looked at your fork but can't find anything about this in any of the branches. **EDIT:** maybe something like this? https://github.com/rust-lang/rust-analyzer/blob/53b6d69e931a52cca44ef94c3b713b4d255754b8/crates/rust-analyzer/src/bin/main.rs#L17-L23 and the doc: https://doc.rust-lang.org/std/alloc/index.html

Yes pretty much. This is the commit I have been running locally since, it doesn't do miracles, but it makes it less worse for little effort.

```diff
From a9190a6970593adf8c9acef34f646fa0dfabd06f Mon Sep 17 00:00:00 2001
From: Jocelyn Turcotte <turcotte.j@gmail.com>
Date: Fri, 15 Apr 2022 23:47:29 +0200
Subject: [PATCH] Use jemalloc to reduce the memory fragmentation with large
 scrollbacks

---
 alacritty/Cargo.toml               | 1 +
 alacritty/src/main.rs              | 5 +++++
 alacritty_terminal/src/grid/row.rs | 2 ++
 3 files changed, 8 insertions(+)

diff --git a/alacritty/Cargo.toml b/alacritty/Cargo.toml
index fd6c19d7..ea501744 100644
--- a/alacritty/Cargo.toml
+++ b/alacritty/Cargo.toml
@@ -39,6 +39,7 @@ unicode-width = "0.1"
 bitflags = "1"
 dirs = "4.0.0"
 once_cell = "1.12"
+tikv-jemallocator = "0.5"

 [build-dependencies]
 gl_generator = "0.14.0"
diff --git a/alacritty/src/main.rs b/alacritty/src/main.rs
index 2e7c30a9..77d3e43f 100644
--- a/alacritty/src/main.rs
+++ b/alacritty/src/main.rs
@@ -62,6 +62,11 @@ use crate::event::{Event, Processor};
 #[cfg(target_os = "macos")]
 use crate::macos::locale;

+use tikv_jemallocator;
+
+#[global_allocator]
+static ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;
+
 fn main() -> Result<(), Box<dyn Error>> {
     #[cfg(windows)]
     panic::attach_handler();
diff --git a/alacritty_terminal/src/grid/row.rs b/alacritty_terminal/src/grid/row.rs
index 2a92a78c..aa7e2237 100644
--- a/alacritty_terminal/src/grid/row.rs
+++ b/alacritty_terminal/src/grid/row.rs
@@ -78,6 +78,8 @@ impl<T: Clone + Default> Row<T> {
         let mut new_row = self.inner.split_off(columns);
         let index = new_row.iter().rposition(|c| !c.is_empty()).map_or(0, |i| i + 1);
         new_row.truncate(index);
+        // Reduce the capacity of the left part.
+        self.inner.shrink_to_fit();

         self.occ = min(self.occ, columns);

--
2.34.1
```

**Comment by correabuscar** [comment]
Thank you both for the info.

I'm reporting the result with this `tikv_jemallocator`: after 6 million lines of history, the `VIRT is 20.9g` and the `RES is 17.6g`, the **window resize** caused it to go to `58.8g VIRT` and `36.0g RES` and stayed there. Closing it via `logout` made the MiB free go from 968 to 45.660. (the `bash` process does keep some `7.6g RES` due to running that command, and 7.9MiB VIRT, so that would be why so much more(7.6g more) got freed).
So I guess it's worse, for me, with jemalloc.

I ran it\* a second time and now it's `17.8g VIRT` and `16.4g RES`, then window resize makes it go to `33.9g VIRT` and `32.6g RES` and \*I've just realized it's because I ran the system/unpatched version, and this is ie. with default system allocator (not jemalloc). So this is the normal way and this is how it looks like normally. (`logout` made it go from 4342MiB free to 45569 MiB, hmm...)

The execution time of the [printf-s](https://github.com/alacritty/alacritty/issues/5438#issuecomment-1289455413) appears to be the same or a bit faster with jemalloc. (like 0.1s-0.5s faster)

I ran the jemalloc version again, I've `20.9g VIRT` and `17.6g RES` after 6 million lines in history, then window resize(shrink horizonal this time, instead of expand horizontal as usual) makes it go to `21.4g VIRT` and `17.6g RES`, then window resize(expand horizontal a bit) makes it go to `58.8g VIRT` and `36.2g RES` (and stays here). `bash` is `7_983_856(bytes) VIRT` and `7.6g RES`. Free mem is 1220 MiB and goes to 46071 MiB after `logout`.

So it's the resize of window on the horizontal where it increases in size that causes the epic mem usage, and apparently not when it's shrinked on the horizontal. (didn't try verticals yet)

I don't know how to use memory profiler yet, but I'm planning to find out. Any ideas welcome? :)

**Comment by kchibisov** [comment]
You'll see the same memory fragmentation from the allocator. You can try `bytehound` or something like that. I'd suggest to retest jemalloc with reasonable amount of scrollback instead of millions.

**Comment by correabuscar** [comment]
I believe I've found the problem (for my case anyway).\* 

The problem is that the Vec doubles its capacity every time it previous capacity *gets filled up and needs 1 more byte*. [This](https://stackoverflow.com/a/40663945/19999437) answer explains it better:
> Indeed, it's notable that fragmentation is not the ONLY source of waste; over-allocation is a common "issue" too:
>
>    a Vec will allocate a power-of-2 number of elements, but maybe you only use 2^N + 1, wasting 2^N - 1 slots

In my case with millions in history, the doubling of the memory when window is being resized by growing it horizontally is easily explained by the above. I've tested it and the following code for example wants 121 columns and gets capacity increase to 240 (and for 6 million lines, doubling the allocated memory this way showed up in 'top' like I've mentioned before):
https://github.com/alacritty/alacritty/blob/e3af53c863ee55b2862954f156e168c1e8da26e2/alacritty_terminal/src/grid/row.rs#L62-L64



(\*but I'm not sure how to use `bytehound` to find this problem exactly(overallocated memory due to doubling of Vec capacity), though `bytehound` definitely help pointing me to the line that was allocating memory)

**Comment by nixpulvis** [comment]
Big yikes, 42 GB is a large number.

```
Process:          alacritty (Alacritty) [7263]
UUID:             677A8E2A-BD75-3035-940F-A640E56CA16E
Path:             /Applications/Alacritty.app/Contents/MacOS/alacritty
Identifier:       org.alacritty
Version:          0.13.0-dev (1)
Is First Party:   No
Shared Cache:     7906534C-5D2B-3BA8-B9C6-93A5CF1B7B50 slid base address 0x1806b8000, slide 0x6b8000 (System Primary)
Architecture:     arm64
Footprint:        42.02 GB
Time Since Fork:  32s
Num samples:      26 (1-26)
CPU Time:         2.498s (4.9G cycles, 11.3G instructions, 0.43c/i)
Note:             1 idle work queue thread omitted
```

Simply ran, `make install` then moved the application to my Applications folder and tried to run the program.

**Comment by Dushistov** [comment]
I just launched the newly installed program from arch linux repository.
And it eats 129M RSS from the start, without any configuration and any commands.

For example gnome-terminal eats 51552 bytes of resident memory,
xterm 29524-45688 bytes. 

![alacritty_rss](https://github.com/alacritty/alacritty/assets/156697/4aea4ac0-2334-4717-ba72-421978ac17e9)

**Comment by kchibisov** [comment]
>And it eats 129M RSS from the start

mesa eats a lot of memory depending on the driver, so I'm not sure what you expect. glxgears which does nothing will each probably 70% of that.

**Comment by Dushistov** [comment]
@kchibisov 

glxgears eats 94180 **bytes** resident memory:
![glxgears_rss](https://github.com/alacritty/alacritty/assets/156697/e926ea54-6502-4ff9-90d5-a07df6c7a3da)

**Comment by kchibisov** [comment]
that's not bytes, it eats 94MB

**Comment by NikoMalik** [comment]
same

**Comment by ghost** [comment]
Just to +1 this, alacritty is using over 200M for me with a default config:

<img width="1011" height="501" alt="Image" src="https://github.com/user-attachments/assets/e1dd6074-c3d7-43f0-86a7-72b9588e0812" />

**Comment by chrisduerr** [comment]
200M Is not unusual at all for some graphics drivers.

**Comment by kchibisov** [comment]
Yeah, just use single instance + multiple windows, extra windows cost nothing pretty much.

**Comment by kgfly** [comment]
About memory consumption: I did some interesting testing across 3 OS, with its competitors.  https://github.com/alacritty/alacritty/issues/8912

The ticket got closed in 9min wo any comments. I guess it's due to its title, lol
